File: day06/day6.js
const fs = require('fs');
const readFile = (path) => fs.readFileSync(path, { encoding: 'utf8' });

const Customs = (file) => {
  const data = readFile(file).split('\n\n');

  const countYesResponses = () => {
    return data.reduce((count, group) => {
      const totalResponses = group.replaceAll('\n', '');
      const uniqueResponses = [...new Set(totalResponses)];
      return count + uniqueResponses.length
    }, 0)
  }

  const countYesForEveryone = () => {
    return data.reduce((count, group) => {
      const responseArray = group.replaceAll('\n', '').split('');
      const numberOfPeople = group.split('\n').length;
      const responseCounts = tally(responseArray);
      Object.values(responseCounts).forEach((total) => {
        if (total === numberOfPeople) count++;
      })
      return count;
    }, 0)
  }

  const tally = (array) => {
    return array.reduce((counterObj, char) => {
      counterObj[char] ? counterObj[char]++ : counterObj[char] = 1;
      return counterObj;
    }, {})
  }

  return {
    countYesResponses,
    countYesForEveryone
  }
}

const customs = Customs(process.argv[2]);

customs.countYesResponses(); // 6735
customs.countYesForEveryone(); // 3221

File: day06/day6.rb
# frozen_string_literal: true

# day6
class Customs
  def initialize(file_data)
    @data = File.read(file_data).split("\n\n")
  end

  def count_yes_responses
    @data.reduce(0) do |count, group|
      count + group.gsub("\n", '').chars.uniq.size
    end
  end

  def count_yes_for_everyone
    @data.reduce(0) do |count, group|
      responses = group.gsub("\n", '').chars.tally
      number_of_people = group.split("\n").size
      responses.each_value { |v| count += 1 if v == number_of_people }
      count
    end
  end
end

customs = Customs.new(*ARGV)

customs.count_yes_responses # 6735
customs.count_yes_for_everyone # 3221
